docs/tutorials/Containers.md

```diff
@@ -0,0 +1,475 @@
+
+#### Configuring Cromwell for Singularity
+
+Once Singularity is installed, the main configuration block you'll need to modify is the `config` block inside `backend.providers` in your Cromwell configuration.
+In particular, this block contains a key called `submit-docker`, which will contain a script that is run whenever a job needs to run that uses a Docker image.
+If the job does not specify a Docker image, then the regular `submit` block will be used instead.
+
+##### Local environments
+
+On local backends, you have to configure Cromwell to use a different `submit-docker` script that would start Singularity instead of docker.
+Singularity requires docker images to be prefixed with the prefix `docker://`.
+An example submit script for Singularity is:
+```bash
+singularity exec --bind ${cwd}:${docker_cwd} docker://${docker} ${job_shell} ${script}
+```
+
+As the Singularity container does not emit a job-id, we must include the `run-in-background` tag within the the provider section in addition to the docker-submit script. As Cromwell watches for the existence of the `rc` file, the `run-in-background` option has the caveat that we require the Singularity container to successfully complete, otherwise the workflow might hang indefinitely.
+
+Putting this together, we have an example base configuration for a local environment:
+```hocon
+include required(classpath("application"))
+
+backend {
+    default: singularity
+    providers: {
+        singularity {
+            # The backend custom configuration.
+            actor-factory = "cromwell.backend.impl.sfs.config.ConfigBackendLifecycleActorFactory"
+
+            config {
+                run-in-background = true
+                runtime-attributes = """
+                  String? docker
+                """
+                submit-docker = """
+                  singularity exec --bind ${cwd}:${docker_cwd} docker://${docker} ${job_shell} ${script}
+                """
+            }
+        }
+    }
+}
+```
+
+##### Job schedulers
+
+At its most basic level, if we want to run Singularity on a job scheduler, we need to pass the singularity command to the scheduler as a wrapped command.
+If we were using SLURM, this means we can use all the normal SLURM configuration as explained in the [SLURM documentation](../backends/SLURM). 
+However, our submit script will be a wrapped version of the script above:
+```
+submit-docker = """
+  # Ensure singularity is loaded if it's installed as a module
+  module load Singularity/3.0.1
+  
+  # Submit the script to SLURM
+  sbatch \
+    --wait \
+    -J ${job_name} \
+    -D ${cwd} \
+    -o ${cwd}/execution/stdout \
+    -e ${cwd}/execution/stderr \
+    -t ${runtime_minutes} \
+    ${"-c " + cpus} \
+    --mem-per-cpu=${requested_memory_mb_per_core} \
+    --wrap "singularity exec --bind ${cwd}:${docker_cwd} docker://${docker} ${job_shell} ${script}"
+  """
+```
```

Comment:
I'm just curious - can you provide an example use case (somewhere in these docs) of when it *is* appropriate to use Singularity on top of Cromwell, and when it isn't? For many users that just need to launch Singularity containers via slurm, Cromwell is overkill. I'm finding that users don't know this - and think they are required to have a workflow manager.

Reply:
My personal use case, I've got workflows with 15-30 tasks each which have a different piece of software (GATK4, bwamem, samtools, etc):
- Each have / I can easily find a docker container for each
- I'm running the workflow on multiple HPCs and cloud environments
- For the life of me I can't figure out how to `module load` through CWL.
- I don't have access to install software, or it's impractical to install the exact version I want
- Reproducibility is a key performance indicator for the workflows I'm running.

Docker has a well established container repository that we can use, but unfortunately we can't run Docker on our HPC. My motivation and seemingly [others](https://github.com/broadinstitute/cromwell/issues/2177#issuecomment-330341279) is, I kind of want a (moderately) drop-in replacement for docker.

___

When it isn't is you don't have a software requirement, you're happy to always module load or always have the modules loaded. eg, the way I'm understanding bcbio, the CWL generated by bcbio talks to the bcbio client, so you don't need to run software in Singularity.

Reply:
I think @vsoch is asking about when you might use purely Singularity on HPC, rather than Cromwell + Singularity. You've mostly been talking about using pure Cromwell.

Reply:
But to answer that question, I think the Cromwell documentation makes it very clear that Cromwell is a WDL + CWL workflow engine (and that you don't need Cromwell unless you have these workflows). I don't think people will find this tutorial without already having decided that they want Cromwell to orchestrate their tasks.

Reply:
I think that makes sense. It seems safe to assume that the reader is already in the Cromwell space. We can keep an eye out for the situation @vsoch is describing and edit as necessary